### PR TITLE
planner: move index advisor into the kernel

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #12303

Problem Summary: planner: move index advisor into the kernel

### What changed and how does it work?

planner: move index advisor into the kernel

1. Add models and a set structure required by the Index Advisor;
2. Add a new interface Optimizer to encapsulate all APIs required by the Index Advisor;
3. Add more UTs

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/optimize.go [hunk 1]
```diff
@@ -623,9 +626,24 @@ func hypoIndexChecker(ctx context.Context, is infoschema.InfoSchema) func(db, tb
...
+	if !ok {
+		return 0, errors.Errorf("plan is not a physical plan: %T", plan)
+	}
+	return core.GetPlanCost(pp, property.RootTaskType, optimizetrace.NewDefaultPlanCostOption())
+}
+
 func init() {
 	core.OptimizeAstNode = Optimize
 	core.IsReadOnly = IsReadOnly
+	indexadvisor.QueryPlanCostHook = queryPlanCost
```
Comment (elsa0520) [inline pkg/planner/optimize.go:646]: Why not implement the interface directly here, but use a hook method to implement it?

Comment (qw4990) [reply pkg/planner/optimize.go:646]: I want to converge as more code as possible into the `indexadvisor` package, and reduce its intrusion to other modules.

File: pkg/planner/indexadvisor/set.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"strings"
+)
+
+// SetKey is the interface for the key of a set item.
+type SetKey interface {
+	Key() string
+}
+
+// Set is the interface for a set.
+type Set[T SetKey] interface {
```
Comment (hawkingrei) [inline pkg/planner/indexadvisor/set.go:29]: If possible, move this code into ```util/set``` to allow more people to use it.

File: pkg/planner/indexadvisor/utils.go [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+					// TODO: log or return this error?
+					continue
+				}
+				possibleColumns = append(possibleColumns, cols...)
+			}
+
+			for _, c := range possibleColumns {
+				colType, err := opt.ColumnType(c)
+				if err != nil {
+					// TODO: log?
```
Comment (hawkingrei) [inline pkg/planner/indexadvisor/utils.go:410]: add log now?

**Comment by qw4990** [comment]
Thanks for your comments, all comments have been addressed! @hawkingrei

**Review by hawkingrei — APPROVED** [review]

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

**Pull request merged**